[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs

PaulZhang12 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150812

Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

PaulZhang12 committed:
FP32 output GEMM from FP16/BF16 inputs

[ghstack-poisoned]

PaulZhang12 committed:
Update on "FP32 output GEMM from FP16/BF16 inputs"




[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1375,6 +1375,16 @@
     SparseCUDA: bmm_out_sparse_cuda
     SparseCsrCUDA: bmm_out_sparse_csr_cuda
 
+- func: bmm.out_dtype(Tensor self, Tensor mat2, ScalarType out_dtype) -> Tensor
```
can you also add optional compute_dtype args?

ngimel commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -1588,6 +1674,44 @@ std::optional<c10::ScalarType> out_dtype) {
+  Tensor out = at::empty({batch1_sizes[0], batch1_sizes[1], batch2_sizes[2]}, batch1.options().dtype(out_dtype));
+  {
+    NoNamesGuard guard;
+    baddbmm_out_cuda_impl(out, out, batch1, batch2, beta, alpha, out_dtype);
```
at this point you no longer need out_dtype argument, you can get dtype from `out`

ngimel commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -1588,6 +1674,44 @@ std::optional<c10::ScalarType> out_dtype) {
+  IntArrayRef batch1_sizes = batch1.sizes();
+  IntArrayRef batch2_sizes = batch2.sizes();
+
+  Tensor out = at::empty({batch1_sizes[0], batch1_sizes[1], batch2_sizes[2]}, batch1.options().dtype(out_dtype));
```
check that input and out dtypes are matching, we support only limited amount for cross-dtype matmuls (e.g. no bf16bf16->fp64 etc)

PaulZhang12 committed:
Update on "FP32 output GEMM from FP16/BF16 inputs"




[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"




[ghstack-poisoned]

PaulZhang12 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -334,8 +344,8 @@ class CuBlasLtMatmulPreference : public CuBlasLtDescriptor<
-template <typename Dtype>
-static inline bool bgemm_internal_cublaslt(CUDABLAS_BGEMM_ARGTYPES(Dtype)) {
+template <typename Dtype, typename C_Dtype = Dtype>
+static inline bool bgemm_internal_cublaslt(CUDABLAS_BGEMM_ARGTYPES_AND_C_DTYPE(Dtype, C_Dtype)) {
```
@ngimel Ah yes missed this function, I figured the unit test would catch any missing ones... Going back to make sure the test can catch this

ngimel commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -617,6 +627,67 @@ void bgemm_internal_cublas<at::Half>(CUDABLAS_BGEMM_ARGTYPES(at::Half)) {
+                                   (void*)alpha_ptr, a, rocblas_datatype_f16_r, (int)lda, stridea,
+                                   b, rocblas_datatype_f16_r, (int)ldb, strideb,
+                                   (void*)beta_ptr, c, rocblas_datatype_f16_r, (int)ldc, stridec,
+                                   c, rocblas_datatype_f16_r, (int)ldc, stridec,
```
seems datatype should be changed here also? Or we should be erroring out earlier for rocm and not have an implementation at all?

PaulZhang12 replied:
Erroring out here is probably best, will do that before

ngimel commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -617,6 +627,67 @@ void bgemm_internal_cublas<at::Half>(CUDABLAS_BGEMM_ARGTYPES(at::Half)) {
 }
 
+template <>
+void bgemm_internal_cublas<at::Half, float>(CUDABLAS_BGEMM_ARGTYPES_AND_C_DTYPE(at::Half, float)) {
```
this looks like a copy paste from `<at::Half>` specialization with the only difference in c dtype, perhaps factor it out as a helper function and set c dtype with std::is_same
```
template <>
void bgemm_internal_cublas<at::Half, at::Half>(CUDABLAS_BGEMM_ARGTYPES(at::Half)) {
  bgemm_internal_cublas_helper<at::Half>(CUDABLAS_BGEMM_ARGS_AND_C_DTYPE(at::Half, at::Half));
}

template <>
void bgemm_internal_cublas<at::Half, float>(CUDABLAS_BGEMM_ARGTYPES_AND_C_DTYPE(at::Half, float)) {
  bgemm_internal_cublas_helper<float>(CUDABLAS_BGEMM_ARGS(at::Half));
}
```

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"




[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"




[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"




[ghstack-poisoned]

eqy commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -617,6 +635,67 @@ void bgemm_internal_cublas<at::Half>(CUDABLAS_BGEMM_ARGTYPES(at::Half)) {
+                                   0, flag)));
+#else
+  cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+  if (prop->major >= 7 && at::globalContext().allowFP16AccumulationCuBLAS()) {
```
Does this codepath make sense as an inclusion? For `half` it's a slightly esoteric option to do accumulation in `half` as well for SKUs that have limited `float` accumulation performance. For `float` output, even if it worked it would functionally be no different than a simple cast IIUC.

ngimel replied:
Good point, I don't know maybe just for API uniformity, even if it doesn't make a practical sense

eqy commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -1067,8 +1226,161 @@ void gemm_internal_cublas<at::BFloat16>(CUDABLAS_GEMM_ARGTYPES(at::BFloat16)) {
+      flag)));
+#else
+  cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+  if (prop->major >= 7 && at::globalContext().allowFP16AccumulationCuBLAS()) {
```
Same situation as above, does it make sense to add output `dtype` as a param when accumulation was done in `half`?

ngimel replied:
I think if cublas itself allows this combination, there's no reason to throw an error for it. If on the other hand cublas doesn't support it then throwing an error seems more reasonable than coming up with workarounds.

PaulZhang12 replied:
@ngimel @eqy will get this combination to throw an error as fp16 accum with fp32 output isn't actually supported in cublas based on docs: https://docs.nvidia.com/cuda/cublas/#cublasgemmstridedbatchedex

eqy replied:
I'm not sure I follow. As @ngimel pointed out, the unfortunately named `allowFP16Accumulation` and `allowBF16Accumulation` flags actually control the math mode, and if they are `true`, we would set e.g., 
`cublasSetMathMode(handle, CUBLAS_DEFAULT_MATH | CUBLAS_MATH_DISALLOW_REDUCED_PRECISION_REDUCTION)` https://docs.nvidia.com/cuda/cublas/#cublassetmathmode. Which I assume since it allows both---would be surprising if it raises a unsupported error.

ngimel replied:
`allowFP16AccumulationCuBLAS` will set compute type to CUBLAS_COMPUTE_16F, and according to the docs, only fp16 Ctype is compatible with that, so we should throw an error if user-requested ctype  is fp32. `allowFP16ReductionCuBLAS` sets MathMode and should be compatible with different ctypes. So I think @PaulZhang12 is on the right track

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"




[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


[ghstack-poisoned]

ngimel commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -556,8 +563,13 @@ void bgemm_internal_cublas<c10::complex<float>>(CUDABLAS_BGEMM_ARGTYPES(c10::com
+void bgemm_internal_cublas_half_helper(CUDABLAS_BGEMM_ARGTYPES_AND_C_DTYPE(at::Half, C_Dtype)) {
+  #ifdef USE_ROCM
+  if (std::is_same_v<C_Dtype, float>)
+    TORCH_CHECK(false, "bgemm input type at::Half and output type float is not supported for ROCm");
```
cc @jeffdaily can someone double check if these combinations are supported in RocBLAS(Lt)?

jeffdaily replied:
Looks like no.  https://rocm.docs.amd.com/projects/hipBLASLt/en/develop/reference/data-type-support.html#supported-data-types

PaulZhang12 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -742,9 +780,45 @@ void bgemm_internal<at::BFloat16>(CUDABLAS_BGEMM_ARGTYPES(at::BFloat16))
+{
+  if (at::globalContext().blasPreferredBackend() == BlasBackend::Cublaslt) {
+    if (!bgemm_internal_cublaslt<at::Half, float>(CUDABLAS_BGEMM_ARGS(at::Half))) {
+      bgemm_internal_cublas<at::Half, float>(CUDABLAS_BGEMM_ARGS(at::Half));
```
Yeah, this was the previous structure of trying cublaslt and then regular cublas...

jeffdaily replied:
That's my doing.  sorry it was confusing.  I'm open to ideas how to improve.  this was to fix a reported issue that hipblaslt failed for certain odd-sized shapes but hipblas worked fine, so we fall back in case of failure.

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eqy commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -1180,8 +1309,50 @@ void gemm_internal<at::BFloat16>(CUDABLAS_GEMM_ARGTYPES(at::BFloat16))
+template<>
+void gemm_internal<at::Half, float>(CUDABLAS_GEMM_ARGTYPES_AND_C_DTYPE(at::Half, float))
+{
+  if (at::globalContext().allowFP16ReductionCuBLAS()) {
```
Is this unsupported or not allowed? Note that this case is _more_ precise than the fp16 accumulation case

PaulZhang12 replied:
I am throwing an error in this case, since we do not want fp16 accumulation -> an fp32 output. Also, does not seem to be supported https://docs.nvidia.com/cuda/cublas/#cublasltmatmul

ngimel replied:
FP16Reduction is not fp16 accumulation, it applies only to split-k algos (results of accumulated block are truncated a few times, 2-4, for the whole gemm). Also, this is a permissive setting (you are allowed to truncate, not you are required to truncate), it likely even won't be used if you have fp32 output, so I'm surprised if it's really throwing an error. Additionally,  allowFP16ReductionCuBLAS=true is default setting in pytorch, so throwing an error in this case is disruptive. Can you double check that it's indeed disallowed in cublasLT?

PaulZhang12 replied:
It should be `allowFP16AccumulationCuBLAS` instead

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

PaulZhang12 commented:
@PaulZhang12 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D73126191).

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

ngimel reviewed (approved):
Thank you! This looks great!

PaulZhang12 committed:
Update on "[CUDA][cuBLAS] Aten GEMM overload for FP32 output from FP16/BF16 inputs"


Enable FP32 output from FP16/BF16 GEMMs in aten with cuBLAS. Accumulation for these GEMMs are generally already done in FP32. Adds the functionality to the following aten operators:
* mm
* bmm
* addmm
* baddmm

Follow up of customer issue: https://github.com/pytorch/pytorch/issues/146241#issuecomment-2781889390 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D73126191](https://our.internmc.facebook.com/intern/diff/D73126191)

[ghstack-poisoned]

PaulZhang12 commented:
@pytorchbot merge

PaulZhang12 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1073,6 +1073,16 @@
     XPU: baddbmm_out_xpu
     SparseCsrCUDA: baddbmm_out_sparse_csr_cuda
 
+- func: baddbmm.dtype(Tensor self, Tensor batch1, Tensor batch2, ScalarType out_dtype, *, Scalar beta=1, Scalar alpha=1) -> Tensor
```
Will do!

ngimel replied:
We should make them public, because a lot of people are asking for them.

ayghri commented:
is this exposed in the python API?
